src/sinks/datadog/logs/sink.rs

```diff
@@ -0,0 +1,376 @@
+    let mut pending_acks: HashMap<u64, usize, BuildHasherDefault<XxHash64>> = HashMap::default();
+    let mut seq_head: u64 = 0;
+    let mut seq_tail: u64 = 0;
+
+    pin!(in_flight);
+
+    loop {
+        gauge!("inflight_requests", in_flight.len() as f64);
+        select! {
+            Some(req) = rx.recv() => {
+                // Rebind the variable to avoid a bug with the pattern matching
+                // in `select!`: https://github.com/tokio-rs/tokio/issues/4076
+                let mut req = req;
+                let seqno = seq_head;
+                seq_head += 1;
+
+                let (tx, rx) = oneshot::channel();
+
+                in_flight.push(rx);
+
+                trace!(
+                    message = "Submitting service request.",
+                    in_flight_requests = in_flight.len()
+                );
+                // TODO: This likely need be parameterized, which builds a
+                // stronger case for following through with the comment
+                // mentioned below.
+                let logic = StdServiceLogic::default();
+                // TODO: I'm not entirely happy with how we're smuggling
+                // batch_size/finalizers this far through, from the finished
+                // batch all the way through to the concrete request type...we
+                // lifted this code from `ServiceSink` directly, but we should
+                // probably treat it like `PartitionBatcher` and shove it into a
+                // single, encapsulated type instead.
+                let batch_size = req.payload_members_len;
+                let finalizers = req.take_finalizers();
+
+                let svc = service.ready().await.expect("should not get error when waiting for svc readiness");
+                let fut = svc.call(req)
+                    .err_into()
+                    .map(move |result| {
+                        logic.update_finalizers(result, finalizers);
+
+                        // If the rx end is dropped we still completed
+                        // the request so this is a weird case that we can
+                        // ignore for now.
+                        let _ = tx.send((seqno, batch_size));
+                    })
+                    .instrument(info_span!("request", request_id = %seqno));
+                tokio::spawn(fut);
+            },
+
+            Some(Ok((seqno, batch_size))) = in_flight.next() => {
+                trace!("pending batch {} finished (n={})", seqno, batch_size);
+                pending_acks.insert(seqno, batch_size);
+
+                let mut num_to_ack = 0;
+                while let Some(ack_size) = pending_acks.remove(&seq_tail) {
+                    num_to_ack += ack_size;
+                    seq_tail += 1
+                }
+                trace!(message = "Acking events.", acking_num = num_to_ack);
+                acker.ack(num_to_ack);
+            },
+
+            else => break
+        }
+    }
+}
```

Comment:
The more I think about this loop the more I think we could do some of it as a `Layer`. There's two kinds of acking going on here, the `Acker` involved one that links up to buffer and then the finalizers that do e2e acking. With the right traits we could probably get this ack concept pulled apart and then allow a layer above retries to call back appropriately. /cc @tobz

Reply:
I'm a little split on making this a `tower` layer.

On the one hand, yes, it could absolutely be made more generic, parameterized, and shared across many, many sinks.  On the other hand, the behavior here is almost entire side effect-ual, rather than meaningfully changing the the response of the call into the service itself.  As is, we literally only care about the response for the purpose of handling finalizers and acking.

I don't think we need to necessarily be hard-and-fast with "layers must modify the response, and cannot simply introduce side effects" but I suppose I'd probably need to see what the idea might look like with pseudocode or something.  Basically, I'm worried we might be making it opaque again, in a slightly different way, for the sake of being able to easily stick it in a `tower` layer vs having it be a standalone reusable component.

Reply:
I'm not loving any of the layer based things I'm coming up with. They work but are they overly specific? Who knows. 

Figure I'll get the tests to go green and leave this pretty well as-is.

Reply:
I agree that the semantics of a layer are maybe not the best fit here, but it seems like it's already _very_ close to being reusable. I think all we'd need to do is make it generic over the request type and we could just reuse this as a function here and in the S3 sink. Could be a good intermediate step at least.

Reply:
Depending on what @vladimir-dd chooses to do in #8929 we'll have three pretty similar sink implementations. It'd be nice to try this conversion on a sink that we know is very different, just to see if this same pattern ends up being useful or not.